src/snowflake/snowpark/_internal/analyzer/cte_utils.py

```diff
@@ -144,16 +149,41 @@ def build_plan_to_query_map_in_post_order(root: "TreeNode") -> None:
         list(duplicate_plan_to_cte_map.values()),
         list(duplicate_plan_to_table_name_map.values()),
     )
-    final_query = with_stmt + SPACE + plan_to_query_map[root]
+    final_query = with_stmt + SPACE + plan_to_query_map[root.encoded_id]
     return final_query
 
 
-def encode_id(
+def encoded_query_id(
     query: str, query_params: Optional[Sequence[Any]] = None
 ) -> Optional[str]:
+    """
+    Encode the query and its query parameter into an id using sha256.
+
+
+    Returns:
+        If encode succeed, return the first 10 encoded value.
+        Otherwise, return None
+    """
     string = f"{query}#{query_params}" if query_params else query
     try:
         return hashlib.sha256(string.encode()).hexdigest()[:10]
     except Exception as ex:
         logging.warning(f"Encode SnowflakePlan ID failed: {ex}")
         return None
+
+
+def encode_id(
+    node_type_name: str, query: str, query_params: Optional[Sequence[Any]] = None
+) -> str:
+    """
+    Encode given query, query parameters and the node type into an id.
+
+    If query and query parameters can be encoded successfully using sha256,
+    return the encoded query id + node_type_name.
+    Otherwise, generate a uuid.
+    """
+    query_id = encoded_query_id(query, query_params)
+    if query_id is not None:
+        return query_id + node_type_name
+    else:
+        return str(uuid.uuid4())
```

Comment:
do we want the same node to have the same `encode_id`. If so, we can rely on `id(node)` which gives you deterministic result, instead of `uuid4()`

Reply:
+1

Reply:
this is used for cached property, so with uuid4 generation it will also be deterministitic, but we can use id also.